### ISSUE TITLE
etchash

### ISSUE THREAD

**Issue by joaogti36** [ethereum-mining/ethminer#2080]
will ethminer have support to the new ETC fork?
will will implement some dag changes  and will be with some "etchash"

**Comment by AndreaLanfranchi** [comment]
As far as I know there are no developers working on this project anymore.
Should someone implement it there's no assurance it will be accomplished within the scheduled hardfork

**Comment by q9f** [comment]
there may be a fork of ethminer available prior to the fork. as for upstream support, there's no certainty given the short deadline. there's a list of supported miners at:

* https://github.com/eth-classic/etchash

**Comment by AndreaLanfranchi** [comment]
As far as I understand from etchash implementation ethminer, as is now, should require no modification to work with Thanos fork.
Assumptions :
- etchash basically will divide by two the epoch height after block x
- as epoch height is a linear search against `seed` provided in each work package ethminer will compute the right size **but will indicate a wrong epoch**

Bottom line ... mining should be fine regardless the fact indicated epoch height will be inaccurate

**Comment by joaogti36** [comment]
there's a fork pool to test... https://mordor.pool.fault.dev/#/ only have ethminer 0.18 in my ethos machine... dunno if will work

**Comment by q9f** [comment]
> As far as I understand from etchash implementation ethminer, as is now, should require no modification to work with Thanos fork.

It needs some miner modifications as it uses the seedhash of the old epoch but the dag of the new epoch from what I understand. 

> there's a fork pool to test... https://mordor.pool.fault.dev/#/ only have ethminer 0.18 in my ethos machine... dunno if will work

it will not work. but let me know if you want to test it. I'm operating that pool.

**Comment by AndreaLanfranchi** [comment]
> It needs some minor modifications as it uses the seedhash of the old epoch but the dag of the new epoch from what I understand.

Can you elaborate this ? Does not make much sense to me.

**Comment by q9f** [comment]
> 
> On `ETCHASH_FORK_BLOCK`, activate the following Etchash logic:
> 
> ```
> const oldEpochLength = 30000
> const newEpochLength = 60000
> const activationBlock = ETCHASH_FORK_BLOCK
> 
> // calcEpochLength returns the epoch length for a given block number.
> func calcEpochLength(blockNum uint64) uint64 {
>   if(blockNum < activationBlock) {
>     return oldEpochLength
>   }
>   return newEpochLength
> }
> 
> // calcEpoch returns the epoch for a given block number
> func calcEpoch(block uint64) int {
>   epochLength := calcEpochLength(block)
>   epoch := int(block / epochLength)
>   return epoch
> }
> 
> // seedHash is the seed to use for generating a verification cache and the
> // mining dataset.
> func seedHash(block uint64) []byte {
>   seed := make([]byte, 32)
>   if block < oldEpochLength {
>     return seed
>   }
>   // keep using oldEpochLength here so seeds don't overlap
>   keccak256 := makeHasher(sha3.NewLegacyKeccak256())
>   for i := 0; i < int(block/oldEpochLength); i++ {
>     keccak256(seed, seed)
>   }
>   return seed
> }
> ```
> The oldEpochLength (30000) changes to newEpochLength (60000) at a given `ETCHASH_FORK_BLOCK` (hardfork required).
> 
> This will result in the epoch (blockNumber/[old/new]EpochLength) halving upon activation. As the DAG cache and dataset sizes are calculated from epoch, the growth in which they have experienced too will half upon activation. DAG caches and dataset will then continue to grow as intended, however at half of the pre-activation rate. To avoid re-use of seeds oldEpochLength will continue to be used within the seedHash function.

Note the _"// keep using oldEpochLength here so seeds don't overlap"_ bit. That's how I understand it.

**Comment by AndreaLanfranchi** [comment]
Ok ... things getting clearer here and, personal opinion, I knew some stupid outcome would happen from a bad design.

Basically reading the code (assuming for example 12M as fork block):
- at `ETCHASH_FORK_BLOCK-1` epoch would have been 399
- at `ETCHASH_FORK_BLOCK` the seed returned is for epoch 400 (as it is computed on oldEpochLength) but miners should build DAG as if it was for epoch 200 (i.e. `round_up(seedEpoch /2)`)

This means miner can acknowledge which DAG (which params to use for its unrolling) **if and only if the blocknumber is also reported in the work package** (as deriving epoch from the seed would simply return the epoch on behalf of oldEpochLength) so 400 instead of 200. So basically **there's a mismatch amongst epoch as reported by seed and the real intended epoch**.

Also `"// keep using oldEpochLength here so seeds don't overlap"` is absolutely nonsense: the seed is not part of the hashing function while the DAG is. So not reusing old seeds makes no sense as in reality etchash is re-using old DAGs. Dumb !

Would have been immensely easier if the ECIP would have considered to tweak DAG params after epoch x for example stating that at epoch 400 those two values changed :
- LightIncrementPerEpoch (from 128K to 64K)
- DagIncrementPerEpoch( from 8MB to 4MB)

This would have pursued the very same result with the following advantages:
- no need to signal block_number on work packages (i.e. less network data)
- consistency amongst seed (as a hash) and epoch number
- minimal changes on miner's side (non convoluted reasoning)
- eventually tweaking the above mentioned parameters to values which **are not** the perfect half of the ethash's ones would also have produced a benevolent effect ... all new generated DAGs would have never seen before on the chain and would void any attempt to reuse already known nonces.

Bottom line is, under the depicted circumnstances, ethminer **won't work for etchash**
Sorry for the rant.

**Comment by vanhalendlrband** [comment]
I hope theres a way to implement i been following your guys AppVeyoe commits and I really love this software. Best miner out there. 

I hope theres a way to get this to work eventually even if after the deadline.

**Comment by planetasrtes** [comment]
Estou precisando entender mais

**Comment by mr-older** [comment]
> Also `"// keep using oldEpochLength here so seeds don't overlap"` is absolutely nonsense: the seed is not part of the hashing function while the DAG is. So not reusing old seeds makes no sense as in reality etchash is re-using old DAGs. Dumb !

```
// Follows Sergio's "STRICT MEMORY HARD HASHING FUNCTIONS" (2014)
// https://bitslog.files.wordpress.com/2013/12/memohash-v0-3.pdf
// SeqMemoHash(s, R, N)
bool static ethash_compute_cache_nodes(
	node* const nodes,
	uint64_t cache_size,
	ethash_h256_t const* seed   <-------
)
{
	uint32_t const num_nodes = (uint32_t) (cache_size / sizeof(node));
	SHA3_512(nodes[0].bytes, (uint8_t*)seed, 32); <-------!!

	for (uint32_t i = 1; i != num_nodes; ++i) {
		SHA3_512(nodes[i].bytes, nodes[i - 1].bytes, 64);
	}
....
}

```
Seed is used in computing cache nodes (many clients are also affected), that's why it was left untouched, probably. But... bad design results in stupid outcome, whatever, that`s right.

Nethertheless, it is not very complex to implement that stupid outcome in the ethminer code. Did it with [https://github.com/ethereum/ethash/tree/master/src/libethash](https://github.com/ethereum/ethash/tree/master/src/libethash), quite easy. The problem is to make it compatible with other chains, possibly requiring new command line arguments and documentation updates. Headache.

**Comment by joaogti36** [comment]
hiveos team released an 19.0.2 fix of ethminer to mine etc after this fork :X where to find it avaiable to the rest of the ppl????

**Comment by vanhalendlrband** [comment]
Cant we just merge this? Its already been done by etclabs, forked from this ethminer repo as well. 

Lets merge this if we can

https://github.com/etclabscore/ethminer

**Comment by joaogti36** [comment]
ethminer 0.19.02 from hiveos team needs at least cuda 10/11 to work with this fork?!?!? i don't see any cuda 8/9 in their repo :X

**Comment by q9f** [comment]
> Cant we just merge this? Its already been done by etclabs, forked from this ethminer repo as well.

It's not mergable yet.

**Comment by unsolvent** [comment]
> https://github.com/etclabscore/ethminer
doesn't work with current epoch 391

**Comment by joaogti36** [comment]
check 0.19.02 from hiveos team... will work nicely.... now if only a cuda8/9 release would be nice to ppl with old drivers aka: 410.*
running ethos....

**Comment by unsolvent** [comment]
> check 0.19.02 from hiveos team... will work nicely.... now if only a cuda8/9 release would be nice to ppl with old drivers aka: 410.*
> running ethos....

hiveos? no thanks. pre-compiled version. Who knows what else they put in the code?

**Comment by unsolvent** [comment]
for any interested...
https://github.com/etclabscore/ethminer
is mining ETC with ethermine again. I think it was the pool who had problems.

**Comment by unsolvent** [comment]
i made a fork "Aquilex" with a watchdog if you guys are interested.
https://youtu.be/GW8jmWBht2c

**Comment by joaogti36** [comment]
> 
> 
> for any interested...
> https://github.com/etclabscore/ethminer
> is mining ETC with ethermine again. I think it was the pool who had problems.

where can we get a release of that... for cuda 8/9 cards?

**Comment by unsolvent** [comment]
> > for any interested...
> > https://github.com/etclabscore/ethminer
> > is mining ETC with ethermine again. I think it was the pool who had problems.
> 
> where can we get a release of that... for cuda 8/9 cards?

Did you try it and see if works?  I know on the code all Cuda are present. 
If Etherminer works this will work too.

**Comment by joaogti36** [comment]
i dont have a compile engine here :( and i need it for ethos :X

**Comment by Ronidemoni** [comment]
Hiveos team has developed a version that works on debian. The file is hive-miners-ethminer-ethash-0.19.0.2-cuda101_0.6-02_all.deb and you can find it in https://download.hiveos.farm/repo/binary/. The next step is to compile it for Windows and other OS.

**Comment by shenwude** [comment]
When ETH2.0 is introduced, I think ETC is the next mining focus. Anyone has plan to develop it? Could do it together:P.
